netbox/resource_netbox_virtual_machine_migrate_v0.go

```diff
@@ -0,0 +1,93 @@
+package netbox
+
+import (
+	"context"
+	"log"
+	"strconv"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
+
+func resourceNetboxVirtualMachineResourceV0() *schema.Resource {
+	return &schema.Resource{
+		Schema: map[string]*schema.Schema{
+			"name": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"cluster_id": &schema.Schema{
+				Type:     schema.TypeInt,
+				Required: true,
+			},
+			"tenant_id": &schema.Schema{
+				Type:     schema.TypeInt,
+				Optional: true,
+			},
+			"platform_id": &schema.Schema{
+				Type:     schema.TypeInt,
+				Optional: true,
+			},
+			"role_id": &schema.Schema{
+				Type:     schema.TypeInt,
+				Optional: true,
+			},
+			"site_id": &schema.Schema{
+				Type:     schema.TypeInt,
+				Computed: true,
+			},
+			"comments": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"memory_mb": &schema.Schema{
+				Type:     schema.TypeInt,
+				Optional: true,
+			},
+			"vcpus": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"disk_size_gb": &schema.Schema{
+				Type:     schema.TypeInt,
+				Optional: true,
+			},
+			"tags": &schema.Schema{
+				Type: schema.TypeSet,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+				Optional: true,
+				Set:      schema.HashString,
+			},
+			"primary_ipv4": &schema.Schema{
+				Type:     schema.TypeInt,
+				Computed: true,
+			},
+		},
+	}
+}
+func resourceNetboxVirtualMachineStateUpgradeV0(_ context.Context, rawState map[string]interface{}, meta interface{}) (map[string]interface{}, error) {
+
+	v, ok := rawState["vcpus"]
+	if !ok {
+		return rawState, nil
+	}
+
+	s, ok := v.(string)
+	if !ok {
+		return rawState, nil
+	}
+
+	log.Printf("[DEBUG] vcpus before migration: %#v", rawState["vcpus"])
+
+	f, err := strconv.ParseFloat(s, 64)
+	if err == nil {
+		rawState["vcpus"] = f
+	} else {
+		rawState["vcpus"] = float64(0)
```

Comment:
Returning an error here would break existing state that contains an empty string.

As for other values that may fail to be parsed:
I think it's best to handle the float parsing gracefully. Otherwise the user is left with an unusable state. 

With the code like this, worst case he has to run `terraform refresh` once.

Reply:
I would assume that the state migration fails and the user has to fix their state to one with a convertible value. Now I'm aware that this probably a very rare case and I think I agree with you.

I just wonder whats the less confusing case for the user: Getting something like "Your state is impossible to migrate. Please change vm.cpus to a float-parsable value" or silently setting the value to an arbitrary value. After all, the log.printf is not echoed unless TF_LOG is set, right?

Reply:
Yes, the log message won't be visible.

I think avoiding manual editing in the state is the way to go. After all it's only a value in the state. It will be overwritten with the real value from netbox on the next apply/refresh anyway.